CBinding/Compiler/ICompiler.cs

```diff
@@ -43,34 +45,152 @@ public interface ICompiler
 		string Name {
 			get;
 		}
-		
+
 		Language Language {
 			get;
 		}
-		
+
 		string CompilerCommand {
 			get;
 		}
-		
+
 		bool SupportsCcache {
 			get;
 		}
-		
+
 		bool SupportsPrecompiledHeaders {
 			get;
 		}
-		
+
 		string GetCompilerFlags (Project project, CProjectConfiguration configuration);
-		
+
 		string GetDefineFlags (Project project, CProjectConfiguration configuration);
-		
+
 		BuildResult Compile (
-		    Project project,
-		    ProjectFileCollection projectFiles,
-		    ProjectPackageCollection packages,
-		    CProjectConfiguration configuration,
-		    ProgressMonitor monitor);
-		
+			Project project,
+			ProjectFileCollection projectFiles,
+			ProjectPackageCollection packages,
+			CProjectConfiguration configuration,
+			ProgressMonitor monitor);
+
 		void Clean (ProjectFileCollection projectFiles, CProjectConfiguration configuration, ProgressMonitor monitor);
 	}
-}
+
+	abstract class CMakeToolchain :CCompiler
+	{
+		CMakeToolchain cmakeToolchain;
+		VS15MSBuildToolchain vs15BuildToolchain;
+		MinGW32Toolchain mingw32Toolchain;
+		/// The name of this toolchain
+		string Name { get; }
+
+		///Whether this toolchain is installed.
+		bool IsInstalled { get; }
+
+		/// Use cmake to generate makefiles for this toolchain.
+		public virtual Task<BuildResult> GenerateMakefiles (
+			string projectName, 
+			FilePath outputDirectory, 
+			ProgressMonitor monitor, 
+			string buildConfiguration)
+		{
+			if (buildConfiguration == "msvc")
+			{
+				return (vs15BuildToolchain.GenerateMakefiles (projectName, outputDirectory, monitor));
+			} else {
+				return (mingw32Toolchain.GenerateMakefiles (projectName, outputDirectory, monitor));	//testing
+			}
+		}
+
+		/// Build using the makefiles for this toolchain.
+		public abstract Task<BuildResult> Build (string projectName, FilePath outputDirectory, ProgressMonitor monitor);
+	}	
+
+
+	abstract class VS15MSBuildToolchain : CMakeToolchain
```

Comment:
Yes, this can be removed as there were some properties inherited from abstract base class.
Since, they are no more of use , we can remove abstract.
Actually lots of inheritance and abstract issues arose because i was trying to save that properties that are relevant to old build system .Now they can easily be removed.